kale/embed/_mpca_torch.py

```diff
@@ -0,0 +1,236 @@
+import warnings
+
+import tensorly as tl
+import torch
+import torch.nn as nn
+
+# from tensorly import fold, unfold
+from tensorly.tenalg import multi_mode_dot
+
+from .mpca import _check_dim_shape
+
+tl.set_backend("pytorch")
+
+
+class MPCA(nn.Module):
+    """MPCA implementation compatible with pytorch
+
+    Args:
+        var_ratio (float, optional): Percentage of variance explained
+            (between 0 and 1). Defaults to 0.97.
+        max_iter (int, optional): Maximum number of iteration. Defaults to 1.
+        mean_remove (bool, optional): Whether remove the mean of input data. Defaults ot True.
+            Set to False if the input data mean is zero already .
+
+
+    Attributes:
+        factors (list of array-like): A list of transposed projection matrices, shapes (P_1, I_1), ...,
+            (P_N, I_N), where P_1, ..., P_N are output tensor shape for each sample.
+        idx_order (tensor): The ordering index of projected (and vectorised) features in decreasing variance.
+        mean_ (tensor): Per-feature empirical mean, estimated from the training set, shape (I_1, I_2, ..., I_N).
+        shape_in (tuple): Input tensor shapes, i.e. (I_1, I_2, ..., I_N).
+        shape_out (tuple): Output tensor shapes, i.e. (P_1, P_2, ..., P_N).
+    """
+
+    def __init__(self, var_ratio=0.97, max_iter=1, mean_remove=True):
+        super().__init__()
+
+        self.var_ratio = var_ratio
+        if max_iter > 0 and isinstance(max_iter, int):
+            self.max_iter = max_iter
+        else:
+            error_msg = "Number of max iterations must be an positive integer but given %s" % max_iter
+            logging.error(error_msg)
+            raise ValueError(error_msg)
+        self.factors = []
+        self.shape_in = ()
+        self.shape_out = ()
+        self.mean_ = 0
+        self.idx_order = None
+        self.n_dim = None
+        self.mean_remove = mean_remove
+
+    def _fit(self, x):
+        """Learn MPCA bases"""
+        shape_ = x.shape  # shape of input data
+        n_spl = shape_[0]  # number of samples
+        n_dim = x.ndim
+        self.n_dim = n_dim
+
+        if self.mean_remove:
+            self.mean_ = torch.mean(x, 0)
+        else:
+            self.mean_ = torch.zeros(x[0, :].shape)
+        x = x - self.mean_
+
+        phi = dict()
+
+        self.shape_in = shape_[1:]
+        shape_out = ()
+
+        for i in range(1, n_dim):
+            for j in range(n_spl):
+                # unfold the j_th tensor along the i_th mode
+                x_ji = tl.unfold(x[j, :], mode=i - 1)
+                if i not in phi:
+                    phi[i] = 0
+                phi[i] = phi[i] + torch.matmul(x_ji, x_ji.T)
+
+        for i in range(1, n_dim):
+            eig_values, eig_vectors = torch.eig(phi[i], eigenvectors=True)
+            eig_values = eig_values[:, 0]
+            idx_sorted = (-1 * eig_values).argsort()
+            # ev_sorted[i] = eig_vectors[:, idx_sorted]
```

Comment:
What for?

Reply:
If it has a purpose, explain with a comment.

Reply:
pytorch `eig` returns eigenvalues in two columns, where the first colum is the real part and the second column is the complex part. Will have explaination in comments.